compiler/src/iree/compiler/Codegen/Common/VectorLayoutAnalysis.cpp

```diff
@@ -0,0 +1,931 @@
+    return ChangeResult::NoChange;
+  case ResolutionResult::Change:
+    return ChangeResult::Change;
+  case ResolutionResult::Conflict: {
+    llvm::errs() << "Layout conflict at: " << *this << "\n";
+    llvm::errs() << "With: " << rhs << "\n";
+    llvm_unreachable("Layout conflict should have been handled with "
+                     "resolveWithPossibleConflict instead");
+  }
+  }
+}
+
+ChangeResult DistributionLayout::resolve(const DistributionLayout *rhs) {
+  assert(rhs && "layout to resolve with should not be null");
+  return resolve(rhs->vectorLayout);
+}
+
+void DistributionLayout::print(raw_ostream &os) const {
+  if (vectorLayout) {
+    os << " " << vectorLayout;
+  } else {
+    os << "Uninitialized";
+  }
+}
+
+void DistributionLayout::onUpdate(DataFlowSolver *solver) const {
+  AnalysisState::onUpdate(solver);
+
+  Value value = point.get<Value>();
+
+  if (propagation) {
+    // Make propagation run again on all users of this value.
+    for (Operation *user : value.getUsers()) {
+      solver->enqueue({user, propagation});
+    }
+    // TODO: Maybe we need to run it on the parent operation as well to give
+    // layout to other results? Seems unlikely though as results usually
+    // don't need the same layout?
+  }
+
+  if (enforcement) {
+    // Make enforcement run on the parent.
+    if (Operation *definingOp = value.getDefiningOp()) {
+      solver->enqueue({definingOp, enforcement});
+    } else {
+      // TODO: This is not always correct. Ideally, we should enqueue all
+      // predecessors of these block arguements.
+      solver->enqueue({value.getParentBlock()->getParentOp(), enforcement});
+    }
+
+    // Enforce users of this value also, as some other operands may need to
+    // be updated.
+    for (Operation *user : value.getUsers()) {
+      solver->enqueue({user, enforcement});
+    }
+  }
+}
+
+/// ==========================================================================
+///        TRANSFER FUNCTIONS UTILITIES
+/// ==========================================================================
+
+/// Get OpOperand from an operation and the lattice index, which is basically
+/// the x^th operand of vector type.
+static OpOperand &getOpOperand(Operation *op, unsigned operandLatticeIndex) {
+  unsigned operandIndex = 0;
+  for (OpOperand &operand : op->getOpOperands()) {
+    if (operand.get().getType().isa<VectorType>()) {
+      if (operandIndex == operandLatticeIndex) {
+        return operand;
+      }
+      operandIndex++;
+    }
+  }
+  llvm_unreachable("No vector operand found");
+}
+
+/// Get a layout if all the given layouts are same. If all layouts are not same,
+/// return nullptr.
+static DistributionLayout *
+getAgreedLayout(ArrayRef<DistributionLayout *> layouts) {
+  if (layouts.size() == 0)
+    return nullptr;
+
+  // Check if all layouts are same.
+  for (unsigned i = 1, e = layouts.size(); i < e; ++i) {
+    if (*layouts[i] != *layouts[0]) {
+      return nullptr;
+    }
+  }
```

Comment:
we have a helper for this:
```suggestion
  if (!llvm::all_equal(llvm::map_range(layouts, [](auto *l) { return *l; })))
    return nullptr;
```